src/HashProtocol.js

```diff
@@ -0,0 +1,82 @@
+import off from 'dom-helpers/events/off';
+import on from 'dom-helpers/events/on';
+import invariant from 'invariant';
+
+import StateStorage from './StateStorage';
+import createPath from './utils/createPath';
+import ensureLocation from './utils/ensureLocation';
+
+export default class HashProtocol {
+  constructor() {
+    this.stateStorage = new StateStorage(this, '@@farce');
+
+    this._index = null;
+    this._numExpectedHashChanges = 0;
+  }
+
+  init() {
+    // TODO: Do we still need to work around the old Firefox bug here?
```

Comment:
This was a well-known Firefox bug, but I don't see it any more. Is it worth keeping the old workaround?